### PR TITLE
Make GCAllocator usable in pure functions

### PR DESCRIPTION
**Pull request by nordlow**
Qualify members of `GCAllocator` as `const` to make it usable in pure contexts.

Because `new` can be used in pure contexts I see no reason why `GCAllocator` shouldn't be that aswell.

### CODE REVIEW

File: std/experimental/allocator/gc_allocator.d [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Written in the D programming language.
 /**
-D's built-in garbage-collected allocator.
+   D's built-in garbage-collected allocator.
```
Comment (wilzbach) [inline std/experimental/allocator/gc_allocator.d]: Unrelated

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: Fixed.

**Comment by nordlow** [comment]
Had to update a `static assert` aswell.

File: std/experimental/allocator/gc_allocator.d [hunk 2]
```diff
@@ -117,31 +117,31 @@ struct GCAllocator
     are `shared`.
     */
 
-    static shared GCAllocator instance;
+    static shared const GCAllocator instance;
```
Comment (n8sh) [inline std/experimental/allocator/gc_allocator.d]: Why not immutable? Does it matter one way or the other?

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: No, AFAICT. An immutable instance is automatically shared, but since GCAllocator contains no data I don't think it matters in this case.

Comment (n8sh) [reply std/experimental/allocator/gc_allocator.d]: That's bizarre.

File: std/experimental/allocator/typed.d [hunk 3]
```diff
@@ -22,7 +22,7 @@ import std.traits : isPointer, hasElaborateDestructor;
 import std.typecons : Flag, Yes, No;
 
 /**
-Allocation-related flags dictated by type characteristics. `TypedAllocator`
+   Allocation-related flags dictated by type characteristics. `TypedAllocator`
```
Comment (n8sh) [inline std/experimental/allocator/typed.d]: No reason for indent

Comment (nordlow) [reply std/experimental/allocator/typed.d]: Fixed.

File: std/experimental/allocator/gc_allocator.d (hunk 2 above)
Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: @andralex @edi33416 ?

**Comment by nordlow** [comment]
Thanks.

**Comment by wilzbach** [comment]
Sorry for being away over the weekend, but I thought we couldn't do this:

> The current code models that reality. Yes, we did make a concession for new to allow pure functions to allocate, but I don't want to use that as precedent to essentially model things that don't exist.

> I suggest we leave the low-level allocators alone. Then use a cast to force makeArray be pure.

https://github.com/dlang/phobos/pull/4733#issuecomment-243412290

(I tried the same ~ two years ago)

**Comment by nordlow** [comment]
Thanks, anyway.

**Comment by wilzbach** [comment]
In case I wasn't clear: I was trying to say that except @andralex changed his mind, I think we need to revert this :/

**Comment by andralex** [comment]
After working on allocators and collection for a good while, I got to the conclusion the allocation funcs must be pure, so let's keep this in.

**Comment by nordlow** [comment]
Great, `MmapAllocator` will be next!

File: std/experimental/allocator/gc_allocator.d (hunk 2 above)
Comment (edi33416) [reply std/experimental/allocator/gc_allocator.d]: The issue seems to be that when you have `static immutable GCAllocator instance` the following
`is(typeof(GCAllocator.instance) == shared)` is `false`, which is odd. 

If you have a look at [Segregator](https://github.com/dlang/phobos/blob/master/std/experimental/allocator/building_blocks/segregator.d#L260) it compares the small and large allocators against `shared` in order to decide how to `mixin` the `Impl` template struct.

With `immutable`, the is expression is `false` and thus the `shared` attribute is not applied.
I don't know why it's false; maybe a compiler bug? @andralex

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: I had the same problems in the static assert. I agree, it looks like a compiler bug.

Comment (edi33416) [reply std/experimental/allocator/gc_allocator.d]: @nordlow @n8sh so this isn't a compiler bug, it was just a misunderstanding.

The is expression `is(T == shared)` checks to see if `T` has `shared` among his type qualifiers, so it is ok to return `false` for `immutable`. `immutable` isn't `shared`, but it is convertible to it.

So the check should be `is(T : shared(T))`. There are a couple of allocators that are affected by this incomplete check; from the top of my head I know of `AffixAllocator` and `Segregator`, though there may be others.

I propose we add a `is(T == immutable)` to those, instead of using `:`, as this is, imho, simpler and clearer. @nordlow do you want to have a go at it?

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: Sure. Ok to add it [this](https://github.com/dlang/phobos/pull/6440)?

Comment (edi33416) [reply std/experimental/allocator/gc_allocator.d]: Probably a new PR would be better

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: Which allocators should we update?

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: Ok, I'll try that.

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: @edi33416, can you have a go at this? My time is currently very spare.

Comment (edi33416) [reply std/experimental/allocator/gc_allocator.d]: @nordlow I'll have at it after DConf

Comment (nordlow) [reply std/experimental/allocator/gc_allocator.d]: Thanks.
